deploy/awstools/awstools.py

```diff
@@ -169,14 +174,20 @@ def launch_instances(instancetype, count, instancemarket, spotinterruptionbehavi
 
     # starting with the first subnet, keep trying until you get the instances you need
     startsubnet = 0
-    instances = []
-    while len(instances) < count:
-        if not (startsubnet < len(subnets)):
-            rootLogger.critical("we tried all subnets, but there was insufficient capacity to launch your instances")
-            rootLogger.critical("""only the following {} instances were launched""".format(len(instances)))
-            rootLogger.critical(instances)
-            return
 
+    if 'fsimcluster' in tags.keys():
+        instances = instances_sorted_by_avail_ip(get_instances_by_tag_type(tags['fsimcluster'], instancetype))
```

Comment:
This is slightly wonky.  There is nothing in the signature of `launch_instances` that requires `fsimcluster` to be a key in `tags` but there may be other tags that should be checked.  Do build hosts use this same `launch_instances` helper or do they have a different one?

Reply:
Also, if we find existing instances in the `fsimcluster`, do we want to check them to make sure they aren't already running `FireSim-f1`?  Might be a good feature but I stopped short of throwing it in with this PR.

Reply:
Yeah, builds use this same helper, see: https://github.com/firesim/firesim/blob/master/deploy/buildtools/buildconfig.py#L38

Reply:
One concern here is if there are workloads that rely on the current "additive" nature of launchrunfarm, i.e. they call it N times asking for one instance each time and expect to end up with N instances after all those calls.

Reply:
Ok.  So, looking at this again, the wonkiness I was referring to is that `fsimcluster` is hardcoded in the logic here. It is also hardcoded in `get_instances_by_tag_type()`.  `get_instances_by_tag_type()` should be generalized to take a dict of tags that gets mapped into a filter in a more generic way instead of having `fsimcluster` hardcoded in either place.

I also need to add another control to address what Sagar calls 'the current "additive" nature of launchrunfarm'.  I think it would be helpful if there was a documented example of a workload that uses that feature because I have trouble understanding it.  I know that's a limitation on my end.

Reply:
@abejgonzalez Do your changes to support other host environments also support growing the runfarm iteratively like this?

Reply:
Oh. I never thought of a run farm being iteratively added to. Frankly, I haven't checked but I presume it should still work for EC2.

Reply:
Yes, but it'd need to be supported by the allocators for other hosts. If we think it's a pain in the ass, then we should put this behavior on a deprecation schedule.

Reply:
Do we know of any workloads that use this additive functionality? If SiFive and Berkeley don't use it then I think its probably safe to deprecate? IMO this functionality isn't clear from the `.ini`s and is tricky to use (to terminate you would have to call terminaterunfarm multiple times to terminate a subset of the instances? to terminate all of them?).

Reply:
> to terminate you would have to call terminaterunfarm multiple times to terminate a subset of the instances? to terminate all of them?

No. The docs are pretty clear that the instance count from the runtime.ini is only used when launching instances.  All other commands use the runfarmtag. https://github.com/firesim/firesim/blame/master/docs/Advanced-Usage/Manager/Manager-Configuration-Files.rst#L50-L52

Regardless, I agree that this should be deprecated somehow to simplify the manager.

Reply:
In the most recent changes that I just pushed, I added an `additive` parameter to `launch_instances` that defaults to `True` and has the old behavior.  If we decide that we don't need, I can take it out and remove the tests for it.